csharp/src/Apache.Arrow/Extensions/StreamExtensions.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Apache.Arrow
+{
+    internal static partial class StreamExtensions
+    {
+        public static async Task EnsureReadFullBufferAsync(this Stream stream, Memory<byte> buffer, CancellationToken cancellationToken = default)
```

Comment:
While there's a ton to love about async/await, one of the downsides from a perf perspective is these little helpers can add measurable allocation cost.  If the any of the reads ends up needing to yield, then this helper that exists purely to check the length and throw will also yield and end up allocating.

Reply:
How can I solve that? Separate the "Ensure" part into a different helper that isn't Async?

```C#
        public static void EnsureFullRead(Memory<byte> buffer, int bytesRead)
        {
            if (bytesRead != buffer.Length)
            {
                throw new InvalidOperationException("Unexpectedly reached the end of the stream before a full buffer was read.");
            }
        }
....

        // caller
        int bytesRead = await BaseStream.ReadFullBufferAsync(buffer).ConfigureAwait(false);
        EnsureFullRead(buffer, bytesRead);
```

Reply:
Yup.